<a href="https://colab.research.google.com/github/Notmrcomedy/Fincomp/blob/main/Taller_2_Computacionales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Juan José Morales

Ana María Duarte

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from scipy.optimize import linprog
import cvxpy as cp

In [19]:
c1 = np.array([3,5])
A1 = np.array([[1,1],
               [2,4],
               [1,2]])
b1 = np.array([500,1000,2000])
bounds1 = [(100,None),
           (150,None)]
res1 = linprog(-c1,A_ub=A1,b_ub=b1,bounds=bounds1,method="revised simplex")
res1

<ipython-input-19-4cbb825fa805>:8: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res1 = linprog(-c1,A_ub=A1,b_ub=b1,bounds=bounds1,method="revised simplex")


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -1350.0
       x: [ 2.000e+02  1.500e+02]
     nit: 2

In [20]:
print(round(pd.DataFrame(res1.x),2)) 

       0
0  200.0
1  150.0


Las cantidades de los productos A y B que maximizan las ganancias de la empresa XYZ  son 200 y 150 respectivamente, que generan una ganancia de 1350 dólares

Punto 2

In [24]:
c2 = np.array([0.1,0.15,0.16,0.08])
A2 = np.array([[1,1,1,1]])
b2 = np.array([1])
D2 = np.array([[0.5,0.3,0.25,0.6],
              [0.3,0.1,0.4,0.2],
              [0.2,0.6,0.35,0.2]])
d2 = np.array([0.35,0.30,0.35])
bounds2 = [(0,None),
           (0,None),
           (0,None),
           (0,None)]
res2 = linprog(-c2,A_eq=A2,b_eq=b2,A_ub=D2,b_ub=d2,bounds=bounds2,method="revised simplex")
res2.fun*-100


<ipython-input-24-1c7962813c09>:12: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res2 = linprog(-c2,A_eq=A2,b_eq=b2,A_ub=D2,b_ub=d2,bounds=bounds2,method="revised simplex")


13.736842105263158

In [ ]:
n2 = 4
x2=cp.Variable(n2)
objetivo2 = cp.Minimize(-c2@x2)
restricciones2=[A2@x2==b2,D2@x2>=d2,x2>=0]
prob = cp.Problem(objetivo2,restricciones2)
prob.solve()
print(round(100*pd.DataFrame(x2.value),2))


       0
0   0.00
1  15.79
2  57.89
3  26.32


Para llegar a un retorno esperado máximo de 13.74%, los pesos asignados a cada uno de los fondos deberían ser los siguientes:
**Fondo 1:** No se invierte
**Fondo 2:** Se invierte un 15.79%
**Fondo 3:** Se invierte un 57.89%
**Fondo 4:** Se invierte un 26.32%

Punto 3


In [30]:
c3 = np.array([0,0,0,0,0,0,0,0,0,0,0,0,0,1])
A3 = np.array([[1,0,0,0,0,1,0,0,-1,0,0,0,0,0],
               [-1.01,1,0,0,0,0,1,0,1.003,-1,0,0,0,0],
               [0,-1.01,1,0,0,0,0,1,0,1.003,-1,0,0,0],
               [0,0,-1.01,1,0,-1.02,0,0,0,0,1.003,-1,0,0],
               [0,0,0,-1.01,1,0,-1.02,0,0,0,0,1.003,-1,0],
               [0,0,0,0,-1.01,0,0,-1.02,0,0,0,0,1.003,-1]
               ])
b3= np.array([150,100,-200,200,-50,-300])
bounds3 = [(0,100),
           (0,100),
           (0,100),
           (0,100),
           (0,100),
           (0,None),
           (0,None),
           (0,None),
           (0,None),
           (0,None),
           (0,None),
           (0,None),
           (0,None),
           (0,None)]
           
res3 = linprog(-c3,A_eq=A3,b_eq=b3,bounds=bounds3,method="revised simplex")
res3.x

<ipython-input-30-9a086b7e7816>:25: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res3 = linprog(-c3,A_eq=A3,b_eq=b3,bounds=bounds3,method="revised simplex")


array([  0.        ,  50.98039216,   0.        ,   0.        ,
         0.        , 150.        ,  49.01960784, 203.43436358,
         0.        ,   0.        , 351.9441675 ,   0.        ,
         0.        ,  92.49694915])

In [34]:
n3=14
x3=cp.Variable(n3)
objetivo3 = cp.Minimize(-c3@x3)
restricciones3=[A3@x3==b3,x3>=0,x3[0:5]<=100]
prob = cp.Problem(objetivo3,restricciones3)
prob.solve(cp.GLPK_MI)
print(round(pd.DataFrame(x3.value),2))
#prob.solve(cp.GLPK_MI)

         0
0    -0.00
1    50.98
2    -0.00
3    -0.00
4    -0.00
5   150.00
6    49.02
7   203.43
8    -0.00
9    -0.00
10  351.94
11   -0.00
12   -0.00
13   92.50


Para maximizar el efectivo en el mes 6, llegando a un valor de 92.5 millones, se prefiere la emisión de bonos sobre el crédito, solo se tiene que solicitar un préstamo por la línea de crédito en el mes 2, de un monto correspondiente a 50,98 millones, lo que significaría un pago de intereses bancarios de 0,5098 millones a lo largo de los 6 meses.

Punto 4

In [37]:
rn = ((0.0002-0.0001)+1)**7
c4 = np.array([0,0,0,0,0,0,0,1])
A4 = np.array([[1,0,0,0,-1,0,0,0],
               [0,1,0,0,rn,-1,0,0],
               [0,0,1,0,0,rn,-1,0],
               [0,0,0,1,0,0,rn,-1]])
b4 = np.array([150,-100,-50,200])
res4 = linprog(-c4,A_eq=A4,b_eq=-b4,method="revised simplex")
-res4.fun

<ipython-input-37-d839a19e0b2d>:8: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res4 = linprog(-c4,A_eq=A4,b_eq=-b4,method="revised simplex")


200.14021366142958

In [36]:
print(round(pd.DataFrame(res4.x),2))

        0
0    0.00
1    0.00
2    0.00
3    0.00
4  150.00
5   50.11
6    0.14
7  200.14


Para llegar a una rentabilidad máxima a final del mes de 200.14 millones, sin dejar de  responder por sus pagos, se debe hacer una inversión en el instrumento de corto plazo de 150 millones en el primer mes, 50.11 millones en el segundo y 0.14 millones en el tercero

---



Punto 5

In [39]:
c5 = np.array([5,8,10,0,0,0])
A5 = ([[1,1,1,0,0,0],
      [0,0,0,1,1,1]])
b5 = np.array([10000,1000])
D5 = np.array([[1,0,0,0,0,0],
               [0,0,0,0,1,0],
               [0,0,0,0,0,1]])

d5 = np.array([2000,300,250])
res5 = linprog(-c5,A_eq=A5,b_eq=b5,A_ub=-D5,b_ub=-d5,method="simplex")
res5


<ipython-input-39-501c5f782d57>:10: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  res5 = linprog(-c5,A_eq=A5,b_eq=b5,A_ub=-D5,b_ub=-d5,method="simplex")


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -90000.0
       x: [ 2.000e+03  0.000e+00  8.000e+03  4.500e+02  3.000e+02
            2.500e+02]
     nit: 3

Para alcanzar una ganancia máxima de 90000$ se le deben asignar 2000 millones de capital y 450 horas de trabajo a  ventas, nada capital pero 300 horas de trabajo a marketing y 8000 millones de capital con 250 horas de trabajo a producción.

Punto 6

In [8]:
c6 = np.array([[4,5,6,8,10],
               [6,4,3,5,8],
               [9,7,4,3,4]])
b6 = np.array([500,500,500])
d6 = np.array([80,270,250,160,180])
x6=cp.Variable((3,5),nonneg=True)
m=np.array([1000,1000,1000])
y6 = cp.Variable(3,boolean=True)
obj6 = cp.sum(cp.multiply(c6,x6))+cp.sum(cp.multiply(m,y6))
objetivo6 = cp.Minimize(obj6)
restricciones6=[cp.sum(x6[0,:])<=b6[0]*y6[0],cp.sum(x6[1,:])<=b6[1]*y6[1],cp.sum(x6[2,:])<=b6[2]*y6[2],
                cp.sum(x6[:,0])==d6[0],cp.sum(x6[:,1])==d6[1],cp.sum(x6[:,2])==d6[2],cp.sum(x6[:,3])==d6[3],cp.sum(x6[:,4])==d6[4],
                cp.sum(y6)==2]
prob = cp.Problem(objetivo6,restricciones6)
prob.solve()
print(round(pd.DataFrame(x6.value),2))


      0      1      2      3      4
0  -0.0   -0.0   -0.0   -0.0   -0.0
1  80.0  270.0  150.0   -0.0   -0.0
2  -0.0   -0.0  100.0  160.0  180.0


Para minimizar sus costos de funcionamiento y transporte la empresa debe en primera instancia cerrar una de las 3 plantas, ya que la cantidad producida por 2, es más que suficiente para suplir la demanda y mantener las 3 abiertas generaría un coste adicional innecesario. En este sentido, se encontró que cerrando la instalación 1 se minimizan los costos de funcionamiento, además para optimizar los costos de transporte se encontró que la mejor forma para satisfacer las demandas de los 5 puntosy minimizar los costos es que la planta 2 produzca 80 unidades para el punto 1, 270 para el punto 2 y 150 para el punto 3, mientras que la planta 3 debería producir 100 unidades para el punto 3, 160 unidades para el punto 4 y 180 para el punto 5